File: cisco-dashboard-frontend/src/types/topology.ts
/**
 * Topology types for Cisco SD-WAN Forensic Dashboard
 * These types mirror the backend models/topology.go structures
 */

/**
 * RelationshipType indicates the source/type of a relationship.
 */
export type RelationshipType = "data-plane" | "control" | "site";

/**
 * TransportSession represents a single BFD/IPsec tunnel within a relationship.
 */
export interface TransportSession {
  color: string;
  state: string;
  srcIp: string;
  dstIp: string;
  proto: string;
  uptime: string;
  uptimeDate?: number;
  txInterval?: number;
  transitions?: number;
  lastUpdated?: number;
}

/**
 * ControlConnection represents a control plane connection detail.
 */
export interface ControlConnection {
  state: string;
  peerType: string;
  protocol: string;
  localColor: string;
  uptime: string;
}

/**
 * SiteLink represents a site topology link detail.
 */
export interface SiteLink {
  linkType: string;
  status: string;
  linkKey: string;
}

/**
 * Relationship represents a logical connection between the selected device and a peer.
 */
export interface Relationship {
  peerIp: string;
  peerHostname: string;
  peerType: string;
  siteId: string | number;
  importance: number;
  healthStatus: "healthy" | "degraded" | "down" | "unknown";
  healthRatio: number;
  relationshipTypes: RelationshipType[];
  transports: TransportSession[];
  controlConns?: ControlConnection[];
  siteLinks?: SiteLink[];
  activeCount: number;
  totalCount: number;
  uniqueColors?: string[];
}

/**
 * HealthSummary provides aggregate health statistics.
 */
export interface HealthSummary {
  healthy: number;
  degraded: number;
  down: number;
  unknown: number;
}

/**
 * LogicalTopologyResponse is the response from /api/topology/logical/{system-ip}
 */
export interface LogicalTopologyResponse {
  selectedDevice: string;
  selectedHostname: string;
  relationships: Relationship[];
  totalPeers: number;
  hiddenCount: number;
  healthSummary: HealthSummary;
}

/**
 * OmpPeer represents an OMP routing peer with route information.
 */
export interface OmpPeer {
  peerIp: string;
  peerHostname: string;
  peerType: string;
  siteId: string;
  routeCount: number;
  vpnIds: string[];
  prefixes: string[];
}

/**
 * OmpTopologyResponse is the response from /api/topology/omp/{system-ip}
 */
export interface OmpTopologyResponse {
  selectedDevice: string;
  selectedHostname: string;
  peers: OmpPeer[];
  totalPeers: number;
  totalRoutes: number;
}

/**
 * Transport / tunnel color map for visualization
 */
export const TRANSPORT_COLORS: Record<string, string> = {
  "public-internet": "#E53935",
  "biz-internet": "#1E88E5",
  "3g": "#3357FF",
  lte: "#FF33A8",
  blue: "#007acc",
  green: "#43A047",
  red: "#FF0000",
  bronze: "#CD7F32",
  silver: "#C0C0C0",
  gold: "#FFD700",
  custom1: "#8A2BE2",
  custom2: "#FF8C00",
  custom3: "#20B2AA",
  mpls: "#F4A460",
  "metro-ethernet": "#8A2BE2",
  private1: "#FFA500",
  private2: "#A52A2A",
  default: "#78909C",
};

/**
 * Get the color for a transport type
 */
export function getTransportColor(transport: string): string {
  return TRANSPORT_COLORS[transport.toLowerCase()] || TRANSPORT_COLORS.default;
}

File: models/topology.go
package models

// RelationshipType indicates the source/type of a relationship.
type RelationshipType string

const (
	RelTypeDataPlane RelationshipType = "data-plane" // From BFD sessions
	RelTypeControl   RelationshipType = "control"    // From control connections
	RelTypeSite      RelationshipType = "site"       // From site topology
)

// TransportSession represents a single BFD/IPsec tunnel within a relationship.
type TransportSession struct {
	Color       string `json:"color"`
	State       string `json:"state"`
	SrcIP       string `json:"srcIp"`
	DstIP       string `json:"dstIp"`
	Proto       string `json:"proto"`
	Uptime      string `json:"uptime"`
	UptimeDate  int64  `json:"uptimeDate,omitempty"`
	TxInterval  int    `json:"txInterval,omitempty"`
	Transitions int    `json:"transitions,omitempty"`
	LastUpdated int64  `json:"lastUpdated,omitempty"`
}

// ControlConnection represents a control plane connection detail.
type ControlConnection struct {
	State      string `json:"state"`
	PeerType   string `json:"peerType"`
	Protocol   string `json:"protocol"`
	LocalColor string `json:"localColor"`
	Uptime     string `json:"uptime"`
}

// SiteLink represents a site topology link detail.
type SiteLink struct {
	LinkType string `json:"linkType"`
	Status   string `json:"status"`
	LinkKey  string `json:"linkKey"`
}

// Relationship represents a logical connection between the selected device and a peer.
type Relationship struct {
	PeerIP            string              `json:"peerIp"`
	PeerHostname      string              `json:"peerHostname"`
	PeerType          string              `json:"peerType"`
	SiteID            interface{}         `json:"siteId"`
	Importance        int                 `json:"importance"`
	HealthStatus      string              `json:"healthStatus"`
	HealthRatio       float64             `json:"healthRatio"`
	RelationshipTypes []RelationshipType  `json:"relationshipTypes"`
	Transports        []TransportSession  `json:"transports"`
	ControlConns      []ControlConnection `json:"controlConns,omitempty"`
	SiteLinks         []SiteLink          `json:"siteLinks,omitempty"`
	ActiveCount       int                 `json:"activeCount"`
	TotalCount        int                 `json:"totalCount"`
	UniqueColors      []string            `json:"uniqueColors,omitempty"`
}

// HealthSummary provides aggregate health statistics.
type HealthSummary struct {
	Healthy  int `json:"healthy"`
	Degraded int `json:"degraded"`
	Down     int `json:"down"`
	Unknown  int `json:"unknown"`
}

// LogicalTopologyResponse is the aggregated topology for a selected device.
type LogicalTopologyResponse struct {
	SelectedDevice   string         `json:"selectedDevice"`
	SelectedHostname string         `json:"selectedHostname"`
	Relationships    []Relationship `json:"relationships"`
	TotalPeers       int            `json:"totalPeers"`
	HiddenCount      int            `json:"hiddenCount"`
	HealthSummary    HealthSummary  `json:"healthSummary"`
}

// OmpPeer represents an OMP routing peer with route information.
type OmpPeer struct {
	PeerIP       string   `json:"peerIp"`
	PeerHostname string   `json:"peerHostname"`
	PeerType     string   `json:"peerType"`
	SiteID       string   `json:"siteId"`
	RouteCount   int      `json:"routeCount"`
	VpnIDs       []string `json:"vpnIds"`
	Prefixes     []string `json:"prefixes"`
}

// OmpTopologyResponse is the OMP routing topology for a device.
type OmpTopologyResponse struct {
	SelectedDevice   string    `json:"selectedDevice"`
	SelectedHostname string    `json:"selectedHostname"`
	Peers            []OmpPeer `json:"peers"`
	TotalPeers       int       `json:"totalPeers"`
	TotalRoutes      int       `json:"totalRoutes"`
}

// BFDSession represents a raw BFD session from vManage.
type BFDSession struct {
	SystemIP        string      `json:"system-ip"`
	SrcIP           string      `json:"src-ip"`
	DstIP           string      `json:"dst-ip"`
	Color           string      `json:"color"`
	LocalColor      string      `json:"local-color"`
	State           string      `json:"state"`
	Proto           string      `json:"proto"`
	Uptime          string      `json:"uptime"`
	UptimeDate      int64       `json:"uptime-date"`
	TxInterval      int         `json:"tx-interval"`
	Transitions     int         `json:"transitions"`
	SiteID          interface{} `json:"site-id"`
	VdeviceHostName string      `json:"vdevice-host-name"`
	VdeviceName     string      `json:"vdevice-name"`
	LastUpdated     int64       `json:"lastupdated"`
}
